Do not allow tutors to override their own assessments if there is a complaint

anditurdiu opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with good performance and prevented too many database calls
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
Fixes #1922
**Edit:** We want to allow the tutor to still see the complaint. However, he/she should not obtain the lock or be able to override the result.

### Description
- [x] Remove transactional & refactor getAllModelingSubmissionsByTutorForExercise
- [x] Remove transactional & refactor getAllTextSubmissionsByTutorForExercise
- [x] Remove transactional & refactor getAllFileUploadSubmissionsByTutorForExercise
- [x] Delete obsolete DB call from ResultRepository and move to SubmissionRepository
- [x] Add a text which lets the tutor know why he cannot override the assessment.
- [x] Make all input elements readOnly, if the tutors assessment has a complaint

### Steps for Testing
1. Create an exam
2. participate and submit it as a student
3. Assess an exercise as a tutor
4. Make a complaint about the assessment as a student
5. As the tutor navigate back to tutor dashboard: You should still see the submission on under your assessed submissions. But you should not be able to obtain the lock and override the assessment. Furthermore, all input fields should be read-only. **Caution: Instructors are always allowed to override**
6. Log in as a different tutor, you should see the submission in your complaints table of the tutor exercise dashboard. You should be able to handle the complaint. The submissions stays in the complaints table also after the complaint has been handled.

### Screenshots
![image](https://user-images.githubusercontent.com/15110960/88105577-6f43a280-cba4-11ea-92af-fc65c2c358e5.png)

krusche commented:
I already quickly talked to @anditurdiu yesterday and I am not really happy with the implementation.
I think tutors should always be able to see their own assessments (e.g. when they want to look up something) and they also should be able to see the complaints about those, to learn from them (of course the student identity should be hidden). 

I think we should solve the issue differently. Instead of hiding the assessment and the complaint, we should rather set it to readonly and make sure the tutor does **not** get the lock when opening the page and is not able to change the assessment or to work on the complaint. This read-only mode should displayed to the tutor with an info text (e.g. at the top or directly above the complaint) like e.g. "This is a complaint about your assessment. Another tutor has to evaluate it so it is read only for you!" (or some similar phrasing)

anditurdiu commented:
> I already quickly talked to @anditurdiu yesterday and I am not really happy with the implementation.
> I think tutors should always be able to see their own assessments (e.g. when they want to look up something) and they also should be able to see the complaints about those, to learn from them (of course the student identity should be hidden).
> 
> I think we should solve the issue differently. Instead of hiding the assessment and the complaint, we should rather set it to readonly and make sure the tutor does **not** get the lock when opening the page and is not able to change the assessment or to work on the complaint. This read-only mode should displayed to the tutor with an info text (e.g. at the top or directly above the complaint) like e.g. "This is a complaint about your assessment. Another tutor has to evaluate it so it is read only for you!" (or some similar phrasing)

I reversed the filtering of submissions with complaints. Now the tutor can see the submission. However, he cannot override it anymore if there is a complaint. This is displayed to the user: **There is a complaint for this assessment. Another tutor must respond.**

anditurdiu commented:
> Tested successfully on test server 1, noticed the below behaviors, which look minor, hence approving.
> 
> **scenario**: tutor 1 assessed original, tutor 2 can evaluate complaint, however cannot override assessment
> Steps:
> 
> 1. as tutor 1 assess original exercise
> 2. as a student, submit a complaint
> 3. as tutor 2, view the complaint
> 4. as tutor 2, try to override the assessment -> actual: cannot override. expected: able to override (clicking the button doesn't do anything.
>    ![attempt to override](https://user-images.githubusercontent.com/63906424/88156293-85d01500-cc09-11ea-8ed2-8c29c7503508.gif)
> 
> **scenario**: tutor 1 assessed original, tutor 2 evaluates complaint, instructor override display text incorrect (original accessor and complaint status) - this might be as expected.
> -> expected: instructor override display text should display tutor 1 as the original accessor. Complaint status should show as resolved. Maybe this is as expected though, the original accessor's exercise display list is also updated accordingly.
> ![image](https://user-images.githubusercontent.com/63906424/88155404-55d44200-cc08-11ea-8386-15d6b345c263.png)
> ![original accessor](https://user-images.githubusercontent.com/63906424/88155901-ffb3ce80-cc08-11ea-8f75-4ed182806506.gif)
> 
> **scenario**: tutor 1 assessed original, tutor 1 opens assessment, display text says they have lock - but tutor 2 can still evaluate the complaint.
> -> expected: display text should not say tutor 1 has the lock, b/c they really don't.
> 
> Others:
> 
> * confirmed original tutor cannot evaluate own complaints
> * confirmed original tutor assessments display as expected before/after complaints - stays on original accessor's list
> * confirmed instructor can override assessments
> 
> ![image](https://user-images.githubusercontent.com/63906424/88155512-76040100-cc08-11ea-9c0a-cba2bf84a9b5.png)

Thank you for the very detailed testing! 
I think there is a slight mixup regarding the wording I used in my description. Overriding a result is not the same as evaluating a complaint. Tutors, who are evaluating the complaints for the assessment of another tutor, should not be able to override the original result. That is why clicking the override button did nothing, as the server was rejecting the request. I updated the header now to hide the button in this case. I also fixed the display issues mentioned in the other scenarios. 
One thing to note: If the instructor **overrides** the assessment, he is placed as the original assessor. This is by design :)

kloessst commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:
```diff
@@ -238,28 +234,16 @@ public TextSubmission save(TextSubmission textSubmission) {
-            return textSubmission;
-        }).collect(toList());
+    public List<TextSubmission> getAllTextSubmissionsAssessedByTutorWithForExercise(Long exerciseId, User tutor) {
+        // Retrieve all submissions assessed by the tutor and filter out ones which contain complaints
```
```suggestion
```

kloessst commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -17,15 +17,39 @@ <h3 class="top-container">
-        <span id="assessmentLockedCurrentUser" *ngIf="isAssessor" class="text-danger ml-3" style="font-size: 65%;" jhiTranslate="artemisApp.assessment.assessmentLockedCurrentUser">
+        <span
+            id="assessmentLockedCurrentUser"
+            *ngIf="(isAssessor && !hasComplaint) || (isAssessor && isAtLeastInstructor)"
```
```suggestion
            *ngIf="isAssessor && (!hasComplaint || isAtLeastInstructor)"
```

kloessst commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -428,14 +427,39 @@ export class FileUploadAssessmentComponent implements OnInit, AfterViewInit, OnD
+        if (this.exercise) {
+            this.isAtLeastInstructor = this.accountService.isAtLeastInstructorInCourse(this.exercise.course || this.exercise.exerciseGroup!.exam!.course);
+        } else {
+            this.isAtLeastInstructor = this.accountService.hasAnyAuthorityDirect(['ROLE_INSTRUCTOR', 'ROLE_ADMIN']);
```
Why do we need the else case here? I don't think we should award everyone with 'ROLE_INSTRUCTOR' these rights in this case.

kloessst commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -428,14 +427,39 @@ export class FileUploadAssessmentComponent implements OnInit, AfterViewInit, OnD
+     * Tutors can override their own results within the assessment due date, if there is no complaint about their assessment.
+     * They cannot override a result anymore, if there is a complaint. Another tutor must handle the complaint.
+     */
+    get canOverride(): boolean {
```
I also saw this code duplicated somewhere else. Maybe we could create a super class for all exercise type assessment components and define this in there.

kloessst commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -495,6 +519,10 @@ export class FileUploadAssessmentComponent implements OnInit, AfterViewInit, OnD
     }
 
+    get readOnly(): boolean {
+        return !this.isAtLeastInstructor && !!this.complaint && this.isAssessor;
```
This appears multiple times. Would be better if this is defined in one place.

anditurdiu replied:
There are three assessment components which have very similar code. However, refactoring them to create an assessment superclass would require a deep intrusion into the code which I feel should be done in a separate PR. This change should also include adapting the manual assessment review of programming exercises to follow the same procedure as the other exercises types. However, this is all out of scope of this PR, and we should probably do this after the exams are over.
